Make Armament as overridable as possible

reaperrr opened the pull request:
To allow downstream mods to ship customised `Armament` derivatives without breaking backwards-compatibility with `Common`.

reaperrr committed:
Make Armament elements as overridable as possible

To allow downstream mods to ship customised Armament-derivatives.

WolfGaming commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -264,7 +264,7 @@ public Barrel CheckFire(Actor self, IFacing facing, Target target)
 		public bool ShouldExplode(Actor self) { return !IsReloading; }
```
You don't want this to be virtual too?

GraionDilach replied:
That's coming from an interface tbh.

Phrohdoh replied:
Could this be farmed off to a public virtual method that the (explicit) interface impl wraps?

GraionDilach commented:
I'm fine with this, but may you link your use-case example, please?

reaperrr commented:
> I'm fine with this, but may you link your use-case example, please?

I can't link anything playable for now (I'd have to upload the content somewhere, which I want to avoid as long as possible), but I could push the code later.

Essentially, I'm using a custom `WeaponSlot` derived from `Armament` to do a few extra things,
- linking weapons to a specific `HeatPool` trait to which they contribute heat per burst
- implementing customisable `TargetPositions` in a way that is good enough for me, without having to wait months until that maybe makes its way into OpenRA

GraionDilach commented:
Was only interested in the code example. The first reasons sounds enough for me. :+1:

reaperrr commented:
Updated.
- made `ShouldExplode` interface explicit
- made `IsReloading` and `OutOfAmmo` virtual
- added virtual `AllowExplode`, which `ShouldExplode` wraps

abcdefg30 commented:
Looks good to me. :+1: / :white_check_mark:

@GraionDilach your 👍 still stands after the recent changes?

GraionDilach commented:
Yeah, I think so.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/1bee32d04a775223189311c4a32fd0227433bd0f...d35f606b0a72147035b497264e7ad693d93d6398)
